File: course_scheduling.py
#Approach: BFS - tOPOLOGICAL SORT
#Time Complexity : O(V + E) WHERE V is courses and E is dependencies
#Space Complexity : O(V + E)

from collections import deque
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        indegree = [0]*numCourses
        pre_map = {}
        
        if prerequisites ==[] or len(prerequisites) == 0: return True
        for edge in prerequisites:
            indegree[edge[0]]+=1
            
            if edge[1] not in pre_map.keys():
                pre_map[edge[1]] = list()
            
            pre_map[edge[1]].append(edge[0])
        
        queue = deque()
        count = 0
        
        for i in range(len(indegree)):
            if indegree[i] == 0:
                queue.append(i)
                count += 1
                
        if(len(queue) == 0): return False
        
        while(len(queue) != 0):
            key = queue.popleft()
            if key in pre_map.keys():
                edges = pre_map[key]
                if edges != None:
                    for x in edges:
                        indegree[x]-=1
                        if indegree[x] == 0:
                            queue.append(x)
                            count+=1
                            if count == numCourses:
                                return True
        return False

File: coursescheduling.java
//TC:O(V+E)
//SC:O(V+E)
//Successfully ran on leetcode

class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites)
    {
        int[] inDegree=new int[numCourses];
        HashMap<Integer,List<Integer>> adjlist=new HashMap<>();


        //Getting inDegree & adjacency list
        for(int[] x: prerequisites)
        {
            int to=x[1];
            int from=x[0];

            inDegree[to]++;

            if(!adjlist.containsKey(from))
            {
                adjlist.put(from, new ArrayList<>());
            }
            adjlist.get(from).add(to);
        }

        //BFS

        Queue<Integer> queue=new LinkedList<>();
        for(int x=0;x<numCourses;x++)
        {
            if(inDegree[x]==0)
                queue.add(x);
        }

        while(!queue.isEmpty())
        {
            int front=queue.poll();

            if(adjlist.containsKey(front))
            {
                for(int x:adjlist.get(front))
            {
                inDegree[x]--;
                if(inDegree[x]==0)
                {
                    queue.add(x);
                }
            }
            }

        }

        //checking for dependencies

        for(int x:inDegree)
        {
            if(x!=0)
            {
                return false;
            }
        }

        return true;
    }
}
